src/Chainweb/Cut.hs

```diff
@@ -207,18 +207,84 @@ cutAdjs c cid = HM.intersection
 -- -------------------------------------------------------------------------- --
 -- Limit Cut Hashes By Height
 
+-- | Find a cut that is a predecessor of the given cut and has a block height
+-- that is smaller or equal the given height.
+--
+-- If the requested limit has larger or equal the current height the given cut
+-- is returned.
+--
+-- Otherwise, the requested height limit is divided by the number of chains
+-- (rounded down) and the result is used such that for each chain the
+-- predecessor of the given cut at the respective height is returned.
+--
 limitCut
     :: HasCallStack
     => WebBlockHeaderDb
     -> BlockHeight
         -- upper bound for the cut height. This is not a tight bound.
     -> Cut
     -> IO Cut
-limitCut wdb h c = c & (cutHeaders . itraverse) f
+limitCut wdb h c
+    | h >= _cutHeight c = return c
+    | otherwise = do
+        c & (cutHeaders . itraverse) fastRoute1
   where
-    ch = h `div` int (order (_chainGraph wdb))
-    f cid x = do
-        db <- give wdb $ getWebBlockHeaderDb cid
+    gorder = order $ _chainGraph wdb
+    ch = h `div` int gorder
+
+    -- if there is only a single block at the requested block height, it must be
+    -- a predecessor of any block of larger height. We first do an efficient
+    -- pointwise lookup and return the result if it is unique. Otherwise we fall
+    -- back to a traversing backward from the block in the given cut.
+    --
+    -- Since we prune databases on startup older parts of the history are
+    -- expected to be linear. Also, generally, there aren't many forks.
+    -- Therefore, we expect this fast code path to be successful in most cases
+    -- where the requested height is further down in the history. If the
+    -- requested height is recent we don't care because the overhead of the
+    -- backward traversal is low.
+    --
+    fastRoute1 cid x = do
+        !db <- give wdb $ getWebBlockHeaderDb cid
+        let l = min (_blockHeight x) (int ch)
+        !a <- S.toList_ & entries db Nothing (Just 2) (Just $ int l) (Just $ int l)
```

Comment:
Does this need to be banged? Won't the pattern match below force the evaluation?

Reply:
yeah, I think, the bang is meaningless here. Also `S.toList_` is strict. If it wasn't, that would be an issue on it's own, because the `entries` stream must not be used outside the `entries` bracket.

Reply:
I wonder how easy it is to leak the stream by using a non-strict accumulator. I guess, since the monad of the stream is fixed to `IO`, any straight forward fold of the stream would be strict in the stream. Not sure, though, if it is also strict in the values in the stream, so that evaluation may still access the underlying rocksdb iterator. If we want to be on the safe side, we should add an `NFData` constraint to the `TreeDB` methods that return streams.